### PR TITLE
Feature 1749 modulefiles

### LINKED ISSUE #1749: Internal: Add modulefiles used for installations on various machines
## Describe the Task ##
Add the modulefiles used for installations on various machines to the repository for reference and backup, to internal/scripts/installation/modulefiles.

### Time Estimate ###
< 1 day of work

### Sub-Issues ###
Consider breaking the task down into sub-issues.
- [ ] *Add a checkbox for each sub-issue here.*

### Relevant Deadlines ###
NONE

### Funding Source ###
2792541

### Assignee ###
- [x] Select **engineer(s)** or **no engineer** required
- [ ] Select **scientist(s)** or **no scientist** required

### Labels ###
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Projects and Milestone ###
- [x] Select **Repository** and/or **Organization** level **Project(s)** or add **alert: NEED PROJECT ASSIGNMENT** label
- [x] Select **Milestone** as the next official version or **Future Versions**

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [x] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdataio](https://github.com/dtcenter/METdataio/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)

## Task Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [x] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [x] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [x] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Add any new Python packages to the [METplus Components Python Requirements](https://metplus.readthedocs.io/en/develop/Users_Guide/overview.html#metplus-components-python-requirements) table.
- [x] Push local changes to GitHub.
- [x] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [x] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Linked issues**
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [x] Iterate until the reviewer(s) accept and merge your changes.
- [x] Delete your fork or branch.
- [x] Close this issue.

### PR DESCRIPTION
**Pull request by jprestop**
## Pull Request Testing ##

- [x] Describe testing already performed for these changes:</br>
None needed

- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>
None needed really. I just wanted to back up these modulefiles for METplus using the same directory structure that we use for the MET modulefiles (and soon to be other components).

- [x] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[None needed]**

- [x] Do these changes include sufficient testing updates? **[None needed]**

- [x] Will this PR result in changes to the test suite? **[No]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>

- [x] Please complete this pull request review by **[20220824]**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [x] Add any new Python packages to the [METplus Components Python Requirements](https://metplus.readthedocs.io/en/develop/Users_Guide/overview.html#metplus-components-python-requirements) table.
- [x] Review the source issue metadata (required labels, projects, and milestone).
- [x] Complete the PR definition above.
- [x] Ensure the PR title matches the feature or bugfix branch name.
- [x] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**
Select: **Organization** level software support **Project** or **Repository** level development cycle **Project**
Select: **Milestone** as the version that will include these changes
- [x] After submitting the PR, select **Development** with the original issue number.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

File: internal/scripts/installation/modulefiles/4.1.1_stampede [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+}
+
+# Could not find a grib_util package on stampede
+module load intel/18.0.2
+module load nco/4.9.3
+module load RstatsPackages/3.5.1
+module load met/10.1.1
+setenv METPLUS_PATH /work2/06612/tg859120/stampede2/METplus/METplus-4.1.1
+# Path to wgrib
+prepend-path PATH /work/07738/kgerheis/stampede2/hpc-stack/v1.0.0-beta1/intel-18.0.2/grib_util/1.2.1/bin
```
Comment (georgemccabe) [inline internal/scripts/installation/modulefiles/4.1.1_stampede:17]: Can wgrib and wgrib2 be obtained via a module load instead? Or is there are reason we are doing it this way? To me it seems like it would be more consistent and maintainable to use module loads.

File: internal/scripts/installation/modulefiles/4.1.1_stampede [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+# Path to wgrib2
+prepend-path PATH /work/07738/kgerheis/stampede2/hpc-stack/v1.0.0-beta1/intel-18.0.2/impi-18.0.2/wgrib2/2.0.8/bin
+prepend-path PATH /work2/06612/tg859120/stampede2/METplus/METplus-4.1.1/ush
```
Comment (georgemccabe) [inline internal/scripts/installation/modulefiles/4.1.1_stampede:20]: I would recommend referencing METPLUS_PATH when setting this path so it is easier to update.

```suggestion
prepend-path PATH ${METPLUS_PATH}/ush
```

**Review by georgemccabe — APPROVED** [review]
These additions look good. I made a few inline suggestions to clean things up. The suggestion for adding the METplus ush directory to the path applies to all of the other modulefiles as well. These changes are not mandatory but I think they would help going forward.

**Comment by jprestop** [comment]
> Can wgrib and wgrib2 be obtained via a module load instead? Or is there are reason we are doing it this way? To me it seems like it would be more consistent and maintainable to use module loads.

Thanks for asking about this.  No, wgrib and wgrib2 cannot currently be obtained via a module load instead.  Unfortunately, when looking at this I discovered that wgrib and wgrib2 can no longer be accessed this way either.

```
login4.stampede2(1059)$ ls /work/07738/kgerheis/stampede2/hpc-stack/v1.0.0-beta1/intel-18.0.2/grib_util/1.2.1/bin
ls: cannot access /work/07738/kgerheis/stampede2/hpc-stack/v1.0.0-beta1/intel-18.0.2/grib_util/1.2.1/bin: 
Permission denied
login4.stampede2(1062)$ ls /work/07738/kgerheis/stampede2/hpc-stack/v1.0.0-beta1/intel-18.0.2/impi-18.0.2/wgrib2/2.0.8/bin
ls: cannot access /work/07738/kgerheis/stampede2/hpc-stack/v1.0.0-beta1/intel-18.0.2/impi-18.0.2/wgrib2/2.0.8/bin:
Permission denied
```

I will write to the stampede helpdesk to ask about having wgrib and wgrib2 installed and accessible with a module load command.

> I would recommend referencing METPLUS_PATH when setting this path so it is easier to update.

I agree that would be ideal, however, it looks like the setting of METPLUS_PATH isn't activated until after the modulefile is loaded.  

I changed:
`prepend-path PATH /work2/06612/tg859120/stampede2/METplus/METplus-4.1.1/ush`
to 
`prepend-path PATH ${METPLUS_PATH}/ush`
and then loaded the METplus modulefile without having previously loaded the METplus modulefile. 

This change produces an error when loading the modulefile and you can see that METPLUS_PATH is also not set due to the error in loading the modulefile:

```
login4.stampede2(1049)$ module load metplus/4.1.1
Lmod has detected the following error:  /work2/06612/tg859120/stampede2/modulefiles/metplus/4.1.1: 
(metplus/4.1.1): can't read "METPLUS_PATH": no such variable 
While processing the following module(s):
    Module fullname  Module Filename
    ---------------  ---------------
    metplus/4.1.1    /work2/06612/tg859120/stampede2/modulefiles/metplus/4.1.1

login4.stampede2(1052)$ echo $METPLUS_PATH

login4.stampede2(1053)$ 
```

Changing:
`prepend-path PATH ${METPLUS_PATH}/ush`
back to 
`prepend-path PATH /work2/06612/tg859120/stampede2/METplus/METplus-4.1.1/ush`

and then loading the modulefile works:
```login4.stampede2(1053)$ module load metplus/4.1.1
login4.stampede2(1054)$ echo $METPLUS_PATH
/work2/06612/tg859120/stampede2/METplus/METplus-4.1.1
```

**Comment by georgemccabe** [comment]
@jprestop, it looks like you can create variables using the `local `keyword to reference elsewhere in a modulefile:

https://lmod.readthedocs.io/en/latest/100_modulefile_examples.html#some-simple-modulefiles

I'm not sure of the exact syntax, but you could try doing something like this:

```
local version_dir = "METplus-4.1.1"
local metplus_path = pathJoin("/work2/06612/tg859120/stampede2/METplus", version_dir)

setenv METPLUS_PATH metplus_path
prepend-path PATH pathJoin(metplus_path, "ush")
```

**Comment by jprestop** [comment]
Thanks for the suggestion!

Unfortunately, that did not work:

```
login3.stampede2(1003)$ module load metplus/4.1.1_test

Lmod is automatically replacing "python2/2.7.15" with "python3/3.7.0".

Lmod has detected the following error:  /work2/06612/tg859120/stampede2/modulefiles/metplus/4.1.1_test: 
(metplus/4.1.1_test): invalid command name "local"
While processing the following module(s):
    Module fullname     Module Filename
    ---------------     ---------------
    metplus/4.1.1_test  /work2/06612/tg859120/stampede2/modulefiles/metplus/4.1.1_test
```

I added to my list as something I can look into in the future as time allows.

**Comment by jprestop** [comment]
@georgemccabe Looking into it briefly, it looks like the syntax in the link you sent is for modulefiles written in lua, whereas these modulefiles are written in tcl.  I've been wanting to look into lua/tcl modulefiles for quite some time, so this is a specific thing to focus on once I do that.
